src/ncp/async_task.cpp

```diff
@@ -0,0 +1,90 @@
+
+/*
+ *  Copyright (c) 2024, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "async_task.hpp"
+
+#include <assert.h>
+#include <memory>
+
+#include "common/code_utils.hpp"
+
+namespace otbr {
+namespace Ncp {
+
+AsyncTask::AsyncTask(const ErrorHandler &aErrorHandler)
+    : mErrorHandler(aErrorHandler)
+{
+}
+
+void AsyncTask::Run(void)
+{
+    SetResult(OT_ERROR_NONE, "");
+}
+
+void AsyncTask::SetResult(otError aError, const std::string &aErrorInfo)
+{
+    if (aError == OT_ERROR_NONE)
+    {
+        if (mThen)
+        {
+            mThen(std::move(mNext));
+        }
+        else if (mErrorHandler)
+        {
+            // The Final case
+            mErrorHandler(aError, aErrorInfo);
+        }
+    }
+    else
+    {
+        if (mErrorHandler)
+        {
+            mErrorHandler(aError, aErrorInfo);
+        }
+    }
+}
```

Comment:
One thought came to me related to this class:

`ThenHandler` receives an `aAsyncTaskPtr` (which is a `shared_ptr`). We use `std::move(mNext)` when calling the `ThenHandler`, so we relinquish our ownership of this `shared_ptr` and pass it to the `ThenHandler`.

We assume the `ThenHandler` will keep the `shared_ptr` alive, and at some point, `SetResult()` will be called on the `aAsyncTaskPtr` instance for it to then proceed with the next chained task and/or report the outcome by calling `ErrorHandler`.

However, what if the `ThenHandler` lets the passed-in `aAsyncTaskPtr` go out of scope (get deleted and freed)? In that case, if I'm not mistaken, we would never call `ErrorHandler`. Is this the desired behavior? I believe we should still call `ErrorHandler` in this scenario to report the outcome (perhaps as failed?).

We could consider handling such a case in the destructor of `AsyncTask`. One thing to consider is that when a chain of linked `AsyncTask` objects is being destroyed, the destructor will be called on each one (`mNext` will be destroyed before `AsyncTask` itself is destroyed), and we need to ensure `ErrorHandler` is called only once (across all chained `AsyncTask`s).

Reply:
Good catch! I think the essence is that we need handle the case where an AsyncTask has ended its lifetime but `SetResult` hasn't been called. 

I added a step to clear the `ResultHandler` when calling `SetResult` because it moves forward anyway and, in the destructor, if `ResultHandler` is not `nullptr`, it's called with a failure. (`SetResult` is never called for this object).

Reply:
Thanks.

However, I believe now we may encounter the issue of the `ResultHandler` potentially being called multiple times. We clear `mResultHandler` when we call it on an `AsyncTask`, but other `AsyncTask`s (in the same chain) will still retain their own copy of the handler in their own `mResultHandler`. When the destructor is called on them, they will also call the same `ResultHandler` again.

Reply:
Here's my suggestion on this:

We modify `AsyncTask` to track either a `ThenHandler` or a `ResultHandler` (not both - we can use a `union`). An `AsyncTask` instance that tracks a `ResultHandler` will always be the end of a chain of tasks. So, if `mNext` is `nullptr`, this is the last entry in the linked list, and therefore it's tracking a `ResultHandler`; otherwise, it's a `ThenHandler`.

```c++
private:
    union
    {
        ThenHandler   mThen;
        ResultHandler mResultHandler;
    };

    AsyncTaskPtr mNext;
```

We'll need to update `First()` and `Then()` to accommodate this new data model.

`SetResult()` would then check if it's not the last entry (has `mNext`) and call `SetResult()` on `mNext`. Otherwise, if it's the last entry, it invokes `ResultHandler` and clears it.

The `~AsyncTask()` destructor can simply call `SetResult()` with the error (in all cases). 

With this model, `ResultHandler` is only invoked from one place.

Thoughts?

Reply:
Thanks a lot! You're correct about this.

I've changed to the union implementation. This actually simplies the logic and saves memory.

The very simple class needs to be taken great care. I also added a unit test `test_async_tasks.cpp` for it.

Reply:
Great. Thank you.